src/probnum/diffeq/odefiltsmooth/ivpfiltsmooth.py

```diff
@@ -166,175 +162,269 @@ def init_implementation(f, y0, t0, prior, initrv, df=None):
+        noisy_component = Normal(
+            mean=np.zeros(current_rv.shape),
+            cov=current_rv.cov.copy(),
+            cov_cholesky=current_rv.cov_cholesky.copy(),
         )
 
-        # 3.2 Update the measurement covariance (measure again)
-        meas_rv, info = self.gfilt.measurement_model.forward_rv(
-            rv=pred_rv, t=t_new, compute_gain=True
+        # Compute the measurements for the error-free component
+        pred_rv_error_free, _ = self.dynamics_model.forward_realization(
+            error_free_state, t=t, dt=dt
+        )
+        meas_rv_error_free, _ = self.measurement_model.forward_rv(
+            pred_rv_error_free, t=t
+        )
+        H = self.measurement_model.linearized_model.state_trans_mat_fun(t=t)
+
+        # Compute the measurements for the full components.
+        # Since the means of noise-free and noisy measurements coincide,
+        # we manually update only the covariance.
+        # The first two are only matrix square-roots and will be turned into proper Cholesky factors below.
+        pred_sqrtm = Phi @ noisy_component.cov_cholesky
+        meas_sqrtm = H @ pred_sqrtm
+        full_meas_cov_cholesky = utils.linalg.cholesky_update(
+            meas_rv_error_free.cov_cholesky, meas_sqrtm
+        )
+        full_meas_cov = full_meas_cov_cholesky @ full_meas_cov_cholesky.T
+        meas_rv = Normal(
+            mean=meas_rv_error_free.mean,
+            cov=full_meas_cov,
+            cov_cholesky=full_meas_cov_cholesky,
         )
 
-        # 4. Update
-        zero_data = np.zeros(meas_rv.mean.shape)
-        filt_rv, _ = self.gfilt.measurement_model.backward_realization(
-            zero_data, pred_rv, rv_forwarded=meas_rv, gain=info["gain"]
+        # Estimate local diffusion_model and error
+        diffusion_error_free = self.diffusion_model.calibrate_locally(
+            meas_rv_error_free
+        )
+        diffusion_full = self.diffusion_model.calibrate_locally(meas_rv)
+        diffusion_for_calibration = self.diffusion_model.update_current_information(
+            diffusion_full, diffusion_error_free, t_new
         )
+        local_errors = np.sqrt(diffusion_for_calibration) * np.sqrt(
+            np.diag(meas_rv_error_free.cov)
+        )
+
+        # Either re-predict and re-measure with improved calibration or let the predicted RV catch up.
+        if self._repeat_after_calibration:
+            pred_rv, _ = self.dynamics_model.forward_rv(
+                rv=current_rv, t=t, dt=dt, _diffusion=diffusion_for_calibration
+            )
+            meas_rv, info = self.measurement_model.forward_rv(
+                rv=pred_rv, t=t_new, compute_gain=True
+            )
+            gain = info["gain"]
+        else:
+            # Combine error-free and noisy-component predictions into a full prediction.
+            # (The measurement has been updated already.)
+            full_pred_cov_cholesky = utils.linalg.cholesky_update(
```

Comment:
If this is indeed not tested, it should probably be :)

Reply:
it is part of the constant diffusion model tests. These are tested. I don't know why the test coverage does not reflect that. (The lines below are the basic update schemes and they are marked as not-tested either, but if they were wrong NONE of the tests would pass).